state/indexer/block/kv/kv.go

```diff
@@ -158,35 +167,25 @@ func (idx *BlockerIndexer) Search(ctx context.Context, q *query.Query) ([]int64,
 		skipIndexes = append(skipIndexes, heightIdx)
 	}
 
-	// Extract ranges. If both upper and lower bounds exist, it's better to get
-	// them in order as to not iterate over kvs that are not within range.
-	ranges, rangeIndexes := indexer.LookForRanges(conditions)
-	var heightRanges indexer.QueryRange
 	if len(ranges) > 0 {
 		skipIndexes = append(skipIndexes, rangeIndexes...)
 
 		for _, qr := range ranges {
-			// If we have a query range over height and want to still look for
-			// specific event values we do not want to simply return all
-			// blocks in this height range. We remember the height range info
-			// and pass it on to match() to take into account when processing events.
-			if qr.Key == types.BlockHeightKey && matchEvents {
-				heightRanges = qr
+			if qr.Key == types.BlockHeightKey && matchEvents && !heightInfo.onlyHeightRange {
 				// If the query contains ranges other than the height then we need to treat the height
 				// range when querying the conditions of the other range.
 				// Otherwise we can just return all the blocks within the height range (as there is no
 				// additional constraint on events)
-				if len(ranges)+1 != 2 {
-					continue
-				}
+				continue
+
 			}
 			prefix, err := orderedcode.Append(nil, qr.Key)
 			if err != nil {
 				return nil, fmt.Errorf("failed to create prefix key: %w", err)
 			}
 
 			if !heightsInitialized {
-				filteredHeights, err = idx.matchRange(ctx, qr, prefix, filteredHeights, true, matchEvents)
+				filteredHeights, err = idx.matchRange(ctx, qr, prefix, filteredHeights, true, matchEvents, heightInfo)
```

Comment:
We need to pass the range here because if we have another range (say `foo > 300`) it should return events only if both `foo > 300` and the height of the event is within the height range . This was previously ignored when `match.events = 1`.